dsl/ChalkODMProp.cc

```diff
@@ -169,21 +169,27 @@ vector<unique_ptr<ast::Expression>> ChalkODMProp::replaceDSL(core::MutableContex
                     // error that value is not a symbol
                     auto typeSymbol =
                         ast::MK::UnresolvedConstant(loc, ast::MK::EmptyTree(), core::Names::Constants::Symbol());
-                    stats.emplace_back(ast::MK::Let(lit->loc, move(val), move(typeSymbol)));
+                    ret.nodes.emplace_back(ast::MK::Let(lit->loc, move(val), move(typeSymbol)));
                 }
             }
         }
 
-        if (foreign == nullptr) {
+        auto hasForeign = foreign != nullptr;
+        if (!hasForeign) {
             auto [fk, foreignTree] = ASTUtil::extractHashValue(ctx, *rules, core::Names::foreign());
             foreign = move(foreignTree);
             if (auto body = thunkBody(ctx, foreign.get())) {
                 foreign = std::move(body);
+                hasForeign = true;
             }
         }
+        auto hasDefault = ASTUtil::hasHashValue(ctx, *rules, core::Names::default_());
+        ret.prop.optional = hasDefault || hasForeign || hasComputedBy;
```

Comment:
After a reading of https://github.com/sorbet/sorbet/blob/master/gems/sorbet-runtime/lib/types/props/_props.rb#L31, I think this logic for determining whether an arg to `new` is optional or not is not correct. I think it might be:

```
arg is optional iff
rules[:factory] is present OR
rules[:default] is present OR
rules[:optional] is not false
```

where `rules` is all the stuff after the first argument passed to `prop` (or `const`).

Reply:
Hm. Logic for `computed_by` is in ChalkODMProp.cc but the `computed_by` prop is not mentioned in that doc comment.

Reply:
After IRL with @jez, there's another condition for when the arg is optional: if the type of the arg is nilable.